### ISSUE TITLE
Editor crashes when polling adb in EditorExportPlatformAndroid after Mutex modernization

### ISSUE THREAD

**Issue by volzhs** [godotengine/godot#46203]
**Godot version:**

3.2.4 rc 8651838268

**OS/device including version:**

Kubuntu 20.10
gcc (Ubuntu 10.2.0-13ubuntu1) 10.2.0

**Issue description:**
Editor crashes when opening project, even empty one.

```
libpthread.so.0!__GI___pthread_mutex_lock(pthread_mutex_t * mutex) (pthread_mutex_lock.c:67)
__gthread_mutex_lock(__gthread_mutex_t * __mutex) (/usr/include/x86_64-linux-gnu/c++/10/bits/gthr-default.h:749)
__gthread_recursive_mutex_lock(__gthread_recursive_mutex_t * __mutex) (/usr/include/x86_64-linux-gnu/c++/10/bits/gthr-default.h:811)
std::recursive_mutex::lock(std::recursive_mutex * const this) (/usr/include/c++/10/mutex:106)
MutexImpl<std::recursive_mutex>::lock(const MutexImpl<std::recursive_mutex> * const this) (./core/os/mutex.h:48)
OS_Unix::execute(OS_Unix * const this, const String & p_path, const List<String, DefaultAllocator> & p_arguments, bool p_blocking, OS::ProcessID * r_child_id, String * r_pipe, int * r_exitcode, bool read_stderr, Mutex * p_pipe_mutex) (./drivers/unix/os_unix.cpp:301)
EditorExportPlatformAndroid::_check_for_changes_poll_thread(void * ud) (./platform/android/export/export.cpp:330)
Thread::callback(Thread * p_self, const Thread::Settings & p_settings, Thread::Callback p_callback, void * p_userdata) (./core/os/thread.cpp:68)
std::__invoke_impl<void, void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*>(void (*&&)(Thread *, const Thread::Settings &, void (*)(void *), void *) __f) (/usr/include/c++/10/bits/invoke.h:60)
std::__invoke<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*>(void (*&&)(Thread *, const Thread::Settings &, void (*)(void *), void *) __fn) (/usr/include/c++/10/bits/invoke.h:95)
std::thread::_Invoker<std::tuple<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*> >::_M_invoke<0ul, 1ul, 2ul, 3ul, 4ul>(std::thread::_Invoker<std::tuple<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*> > * const this) (/usr/include/c++/10/thread:264)
std::thread::_Invoker<std::tuple<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*> >::operator()(std::thread::_Invoker<std::tuple<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*> > * const this) (/usr/include/c++/10/thread:271)
std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*> > >::_M_run(std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(Thread*, Thread::Settings const&, void (*)(void*), void*), Thread*, Thread::Settings, void (*)(void*), void*> > > * const this) (/usr/include/c++/10/thread:215)
execute_native_thread_routine (Unknown Source:0)
libpthread.so.0!start_thread(void * arg) (pthread_create.c:463)
libc.so.6!clone() (clone.S:95)
```

and I got `pthread_mutex_lock.c` said below
```
Could not load source './nptl/../nptl/pthread_mutex_lock.c': 'SourceRequest' not supported..
```

**Steps to reproduce:**
1. Run editor
2. Creating a project
3. Open & edit

**Minimal reproduction project:**

**Comment by volzhs** [comment]
reverting 220f24c191 fixed this issue.
cc @RandomShaper @akien-mga

**Comment by qarmin** [comment]
Can't reproduce with Ubuntu 20.04 and Linux Mint 20
Are you using VSCode? If yes this may be an issue - https://github.com/microsoft/vscode-cpptools/issues/3831

**Comment by volzhs** [comment]
I got crash before I was trying to find out what's happening with VSCode.
I compiled editor as same always, but this time I got crash.

**Comment by volzhs** [comment]
I can run editor after removing `export/android/android_sdk_path` from editor setting.
@qarmin did you set this value?

**Comment by qarmin** [comment]
I did not have this option set before.
 
I tried to set this value to random folder and then restart editor, but still I don't have a crash.
Looks that proper sdk directory is needed for it.

**Comment by volzhs** [comment]
I think it is. what I found is that it crashes when trying to access `adb`

**Comment by akien-mga** [comment]
Most likely a case which was missed by @RandomShaper while converting others, or which was added by another PR in parallel.

**Comment by akien-mga** [comment]
It crashes on a simple `OS.execute("ls", ["-l"])`.

```
(gdb) bt
#0  0x00007ffff7d69d50 in pthread_mutex_lock () from /lib64/libpthread.so.0
#1  0x000000000171345d in __gthread_mutex_lock (__mutex=0x0) at /usr/include/c++/10/x86_64-mageia-linux-gnu/bits/gthr-default.h:749
#2  0x000000000171347e in __gthread_recursive_mutex_lock (__mutex=0x0) at /usr/include/c++/10/x86_64-mageia-linux-gnu/bits/gthr-default.h:811
#3  0x00000000017134ce in std::recursive_mutex::lock (this=0x0) at /usr/include/c++/10/mutex:106
#4  0x00000000027eb5c3 in MutexImpl<std::recursive_mutex>::lock (this=0x0) at ./core/os/mutex.h:48
#5  OS_Unix::execute (this=0x7fffffffce70, p_path=..., p_arguments=..., p_blocking=true, r_child_id=0x7fffffffb268, r_pipe=0x7fffffffb250, r_exitcode=0x7fffffffb264, read_stderr=false, p_pipe_mutex=0x0)
    at drivers/unix/os_unix.cpp:301
#6  0x000000000447df83 in _OS::execute (this=0x66e1530, p_path=..., p_arguments=..., p_blocking=true, p_output=..., p_read_stderr=false) at core/bind/core_bind.cpp:521
#7  0x00000000044d79f8 in MethodBind5R<int, String const&, Vector<String> const&, bool, Array, bool>::call (this=0x66e84e0, p_object=0x66e1530, p_args=0x7fffffffb5d8, p_arg_count=2, r_error=...)
    at ./core/method_bind.gen.inc:4177
#8  0x000000000423bdee in Object::call (this=0x66e1530, p_method=..., p_args=0x7fffffffb5d8, p_argcount=2, r_error=...) at core/object.cpp:919
#9  0x00000000042cb9da in Variant::call_ptr (this=0x7a893d8, p_method=..., p_args=0x7fffffffb5d8, p_argcount=2, r_ret=0x0, r_error=...) at core/variant_call.cpp:1129
#10 0x000000000182f85c in GDScriptFunction::call (this=0x7c440e0, p_instance=0x7c93c70, p_args=0x0, p_argcount=0, r_err=..., p_state=0x0) at modules/gdscript/gdscript_function.cpp:1086
#11 0x00000000017d70ce in GDScriptInstance::_ml_call_reversed (this=0x7c93c70, sptr=0x7c94bb0, p_method=..., p_args=0x0, p_argcount=0) at modules/gdscript/gdscript.cpp:1239
#12 0x00000000017d7140 in GDScriptInstance::call_multilevel_reversed (this=0x7c93c70, p_method=..., p_args=0x0, p_argcount=0) at modules/gdscript/gdscript.cpp:1246
#13 0x00000000033e1ab2 in Node::_notification (this=0x7cd8510, p_notification=13) at scene/main/node.cpp:149
#14 0x000000000179b1b0 in Node::_notificationv (this=0x7cd8510, p_notification=13, p_reversed=false) at ./scene/main/node.h:46
#15 0x000000000179b7d5 in CanvasItem::_notificationv (this=0x7cd8510, p_notification=13, p_reversed=false) at ./scene/2d/canvas_item.h:166
#16 0x000000000386a363 in Node2D::_notificationv (this=0x7cd8510, p_notification=13, p_reversed=false) at ./scene/2d/node_2d.h:38
#17 0x00000000039892d9 in Sprite::_notificationv (this=0x7cd8510, p_notification=13, p_reversed=false) at scene/2d/sprite.h:39
#18 0x000000000423bf0b in Object::notification (this=0x7cd8510, p_notification=13, p_reversed=false) at core/object.cpp:929
#19 0x00000000033e1ead in Node::_propagate_ready (this=0x7cd8510) at scene/main/node.cpp:196
#20 0x00000000033e1e34 in Node::_propagate_ready (this=0x7c9ba10) at scene/main/node.cpp:188
#21 0x00000000033e1e34 in Node::_propagate_ready (this=0x7acfab0) at scene/main/node.cpp:188
#22 0x00000000033ee5a0 in Node::_set_tree (this=0x7acfab0, p_tree=0x7acde00) at scene/main/node.cpp:2613
#23 0x0000000003411d0a in SceneTree::init (this=0x7acde00) at scene/main/scene_tree.cpp:463
#24 0x0000000001710e29 in OS_X11::run (this=0x7fffffffce70) at platform/x11/os_x11.cpp:3615
#25 0x00000000016ff72c in main (argc=1, argv=0x7fffffffd768) at platform/x11/godot_x11.cpp:56
```

**Comment by qarmin** [comment]
I have this errors before crash
```
drivers/unix/os_unix.cpp:301:22: runtime error: member call on null pointer of type 'struct Mutex'
core/os/mutex.h:48:3: runtime error: member access within null pointer of type 'const struct MutexImpl'
core/os/mutex.h:48:3: runtime error: member call on null pointer of type 'struct recursive_mutex'
/usr/include/c++/9/mutex:106:47: runtime error: member access within null pointer of type 'struct recursive_mutex'
/usr/include/x86_64-linux-gnu/c++/9/bits/gthr-default.h:749:41: runtime error: null pointer passed as argument 1, which is declared to never be null
handle_crash: Program crashed with signal 11
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
Naruszenie ochrony pamięci (zrzut pamięci)
```

**Comment by akien-mga** [comment]
After #45618 `Mutex` should no longer be passed as a pointer, but it still is in `OS::execute()`. Pedro seemed to start doing some changes in `OS_Unix`'s implementation, but not in `OS_Windows` for example. Not sure how it should be done, a thorough review of all `OS::execute` implementations is needed:
```
$ rg "Mutex \*"
platform/windows/os_windows.cpp
2882:Error OS_Windows::execute(const String &p_path, const List<String> &p_arguments, bool p_blocking, ProcessID *r_child_id, String *r_pipe, int *r_exitcode, bool read_stderr, Mutex *p_pipe_mutex) {

platform/windows/os_windows.h
498:    virtual Error execute(const String &p_path, const List<String> &p_arguments, bool p_blocking = true, ProcessID *r_child_id = NULL, String *r_pipe = NULL, int *r_exitcode = NULL, bool read_stderr = false, Mutex *p_pipe_mutex = NULL);

platform/javascript/os_javascript.cpp
1050:Error OS_JavaScript::execute(const String &p_path, const List<String> &p_arguments, bool p_blocking, ProcessID *r_child_id, String *r_pipe, int *r_exitcode, bool read_stderr, Mutex *p_pipe_mutex) {

platform/javascript/os_javascript.h
168:    virtual Error execute(const String &p_path, const List<String> &p_arguments, bool p_blocking = true, ProcessID *r_child_id = NULL, String *r_pipe = NULL, int *r_exitcode = NULL, bool read_stderr = false, Mutex *p_pipe_mutex = NULL);

platform/uwp/os_uwp.h
208:    virtual Error execute(const String &p_path, const List<String> &p_arguments, bool p_blocking = true, ProcessID *r_child_id = NULL, String *r_pipe = NULL, int *r_exitcode = NULL, bool read_stderr = false, Mutex *p_pipe_mutex = NULL);

platform/uwp/os_uwp.cpp
723:Error OS_UWP::execute(const String &p_path, const List<String> &p_arguments, bool p_blocking, ProcessID *r_child_id, String *r_pipe, int *r_exitcode, bool read_stderr, Mutex *p_pipe_mutex) {

drivers/unix/os_unix.cpp
271:Error OS_Unix::execute(const String &p_path, const List<String> &p_arguments, bool p_blocking, ProcessID *r_child_id, String *r_pipe, int *r_exitcode, bool read_stderr, Mutex *p_pipe_mutex) {

drivers/unix/os_unix.h
88:     virtual Error execute(const String &p_path, const List<String> &p_arguments, bool p_blocking = true, ProcessID *r_child_id = NULL, String *r_pipe = NULL, int *r_exitcode = NULL, bool read_stderr = false, Mutex *p_pipe_mutex = NULL);

core/os/os.h
289:    virtual Error execute(const String &p_path, const List<String> &p_arguments, bool p_blocking = true, ProcessID *r_child_id = NULL, String *r_pipe = NULL, int *r_exitcode = NULL, bool read_stderr = false, Mutex *p_pipe_mutex = NULL) = 0;
```

**Comment by qarmin** [comment]
Looks that OS::execute always passes as the latest argument null because it is not binded 
https://github.com/godotengine/godot/blob/be8b7571c871ef12c4af299bedc4b09a88b84c8c/core/bind/core_bind.cpp#L513-L521
https://github.com/godotengine/godot/blob/be8b7571c871ef12c4af299bedc4b09a88b84c8c/core/bind/core_bind.cpp#L1334

at the end should be a mutex
https://github.com/godotengine/godot/blob/8651838268111b0bbb66c6ab57a0f42869b0e182/drivers/unix/os_unix.cpp#L271

**Comment by qarmin** [comment]
Looks that this patch works(there is memory leak and mutex must be deleted) but I don't know if it is proper
```
diff --git a/core/bind/core_bind.cpp b/core/bind/core_bind.cpp
index 04e668049c..0a746bd226 100644
--- a/core/bind/core_bind.cpp
+++ b/core/bind/core_bind.cpp
@@ -512,13 +512,14 @@ Error _OS::shell_open(String p_uri) {
 
 int _OS::execute(const String &p_path, const Vector<String> &p_arguments, bool p_blocking, Array p_output, bool p_read_stderr) {
 
+    Mutex* mutex =new  Mutex();
        OS::ProcessID pid = -2;
        int exitcode = 0;
        List<String> args;
        for (int i = 0; i < p_arguments.size(); i++)
                args.push_back(p_arguments[i]);
        String pipe;
-       Error err = OS::get_singleton()->execute(p_path, args, p_blocking, &pid, &pipe, &exitcode, p_read_stderr);
+    Error err = OS::get_singleton()->execute(p_path, args, p_blocking, &pid, &pipe, &exitcode, p_read_stderr, mutex);
        p_output.clear();
        p_output.push_back(pipe);
        if (err != OK)
```

**Comment by RandomShaper** [comment]
> After #45618 `Mutex` should no longer be passed as a pointer

Let me clarify that.

While it's true that in many cases we no longer need to use pointers for `Mutex`, `Thread`, `Semaphore` and `RW_Lock`, there are still a few exceptions:
- When it's not known at runtime if the object will indeed be created or not (like in the new resource loader, because it won't create it if not using threaded loading).
- When it's an optional argument.
- When objects of those kinds need to be stored in a container that, because the inability to copy them would prevent it. For instance, you can't have a `Vector<Mutex>` (I think `List<Mutex>` would be OK since it works with pointers internally and doesn't need to copy the objects around).

**Comment by akien-mga** [comment]
Fixed by #46198.
